R/derive_param_clinbenefit.R

```diff
@@ -0,0 +1,266 @@
+#' )
+#'
+#' resp <- date_source(
+#'   dataset_name = "adrs",
+#'   date = ADT,
+#'   filter = PARAMCD == "RSP" & AVALC == "Y" & ANL01FL == "Y"
+#' )
+#'
+#' derive_param_clinbenefit(
+#'   dataset = adrs,
+#'   dataset_adsl = adsl,
+#'   filter_source = PARAMCD == "OVR",
+#'   source_resp = resp,
+#'   source_pd = pd,
+#'   source_datasets = list(adrs = adrs),
+#'   reference_date = TRTSDT,
+#'   ref_start_window = 28,
+#'   set_values_to = vars(
+#'     PARAMCD = "CBR"
+#'   )
+#' )
+derive_param_clinbenefit <- function(dataset,
+                                     dataset_adsl,
+                                     filter_source,
+                                     source_resp,
+                                     source_pd,
+                                     source_datasets,
+                                     reference_date,
+                                     ref_start_window,
+                                     aval_fun = yn_to_numeric,
+                                     set_values_to,
+                                     subject_keys = vars(STUDYID, USUBJID)) {
+
+  # Assertions and quotes
+  reference_date <- assert_symbol(enquo(reference_date))
+  assert_vars(subject_keys)
+  assert_data_frame(
+    dataset_adsl,
+    required_vars = vars(!!!subject_keys, !!reference_date)
+  )
+  assert_data_frame(
+    dataset,
+    required_vars = vars(!!!subject_keys, PARAMCD, AVALC, ADT)
+  )
+
+  filter_source <- assert_filter_cond(
+    enquo(filter_source),
+    optional = FALSE
+  )
+  assert_function(aval_fun)
+  assert_s3_class(source_resp, "date_source")
+  assert_s3_class(source_pd, "date_source")
+  assert_list_of(source_datasets, "data.frame")
+
+  source_names <- names(source_datasets)
+  if (!all(c(source_pd$dataset_name, source_resp$dataset_name) %in% source_names)) {
+    abort(
+      paste0(
+        "The dataset names specified for `source_pd` and `source_resp` must be \n",
+        "included in the list specified for the `source_datasets` parameter.\n",
+        "Following names were provided by `source_datasets`:\n",
+        enumerate(source_names, quote_fun = squote)
+      )
+    )
+  }
+
+  ref_start_window <- assert_integer_scalar(ref_start_window)
+  assert_varval_list(set_values_to, optional = TRUE)
+  assert_param_does_not_exist(dataset, quo_get_expr(set_values_to$PARAMCD))
+
+  # ADSL variables
+
+  adsl_vars <- vars(
+    !!!subject_keys,
+    !!reference_date
+  )
+
+  adsl <- dataset_adsl %>%
+    select(!!!adsl_vars)
+
+  # Get response date
+
+  rsp_data <- source_datasets[[source_resp$dataset_name]] %>%
+    filter_if(source_resp$filter) %>%
+    select(!!!subject_keys, !!source_resp$date) %>%
+    rename(temp_rs = !!source_resp$date)
+
+  # Look for valid non-PD measurements after window from reference date
+  ovr_data <- filter_pd(
+    dataset = dataset,
+    filter = !!filter_source,
+    source_pd = source_pd,
+    source_datasets = source_datasets,
+    subject_keys = vars(!!!subject_keys)
```

Comment:
Please simplify to `subject_keys = subject_keys`.

Reply:
done